pkg/daemon/certs/certs.go

```diff
@@ -0,0 +1,192 @@
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate: %v\n", err)
+		}
+		certs = append(certs, cert)
+	}
+	return certs, nil
+}
+
+// splitAt is a custom closure function to split data into complete PEM-encoded certificates
+func splitAt(substring string) func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+	searchBytes := []byte(substring)
+	searchLen := len(searchBytes)
+	return func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+		dataLen := len(data)
+		if atEOF && dataLen == 0 {
+			// Empty file, do nothing
+			return 0, nil, nil
+		}
+
+		// Find next separator and return the token
+		if i := bytes.Index(data, searchBytes); i >= 0 {
+			return i + searchLen, data[0 : i+searchLen], nil
+		}
+
+		if atEOF {
+			return dataLen, data, fmt.Errorf("Hit end of file without terminating the current token")
+		}
+		// Otherwise, continue reading file data
+		return 0, nil, nil
+	}
+}
+
+// getAllCerts reads all the certs from the given certificate store
+func getAllCerts(store windows.Handle) ([]*x509.Certificate, error) {
+	var existingSystemCerts []*x509.Certificate
+	var certContext *windows.CertContext
+	var err error
+	for {
+		certContext, err = windows.CertEnumCertificatesInStore(store, certContext)
+		if err != nil {
+			if errors.Is(err, windows.Errno(windows.CRYPT_E_NOT_FOUND)) {
+				// Error code implies we have read all certs:
+				// https://learn.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-certenumcertificatesinstore
+				break
+			}
+			return nil, fmt.Errorf("Error reading existing system certificate: %v", err)
+		}
+
+		// Convert the certificate bytes to Golang x509.Certificate
+		certBytes := unsafe.Slice(certContext.EncodedCert, certContext.Length)
+		x509Cert, err := x509.ParseCertificate(certBytes)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to parse certificate from bytes %s: %v", certBytes, err)
+		}
+		existingSystemCerts = append(existingSystemCerts, x509Cert)
+	}
+	klog.Infof("found %d existing system certificates", len(existingSystemCerts))
+	return existingSystemCerts, nil
+}
+
+// ensureCertsAreImported imports any of the expected certificates that are not already present in the existing list
+// into the given Windows certificate store. Returns a boolean if any certificates were imported
+func ensureCertsAreImported(store windows.Handle, expected, existing []*x509.Certificate) (bool, error) {
+	certChange := false
+	for _, cert := range expected {
+		if containsCert(existing, cert) {
+			// Cert already exists as expected, do nothing
+			continue
+		}
+
+		// Convert x509 certificate to a Windows CertContext
+		certContext, err := windows.CertCreateCertificateContext(
+			windows.X509_ASN_ENCODING|windows.PKCS_7_ASN_ENCODING, &cert.Raw[0], uint32(len(cert.Raw)))
+		if err != nil {
+			return false, fmt.Errorf("Failed to create content from x509 cert %s: %v\n", cert.Subject.String(), err)
```

Comment:
Is the subject always populated?

Reply:
It'll always be non-nil. Worst case Subject.String() will return empty string so no worry of panic